File: functions/CopBET_intranetwork_synchrony.m
% out = CopBET_intranetwork_synchrony(in,atlas,varargin)
%
% Copenhagen Brain Entropy Toolbox: Intranetwork synchrony.
% Evaluates intranetwork synchrony as in Carhart-Harris et al., 2014.
% Denoised voxel-wise timeseries are loaded along with a set of ROIs in a
% 4D logical matrix. For each volume, for each roi, the mean os all voxels
% in the ROI is evaluated as well as the variance across voxels. 
% Then, for each ROI, probability mass function
% of the variances over all time points in the
% ROI is established, as well as the Shannon entropy. 
%
% Input:
%   in: char with the path to the input, denoised voxel-wise time series or
%   a table where the first column contains
%   chars (in cells), e.g., different subjects or scan sessions.
%   atlas: The ROIs in a 4D logical matrix. 
% name-value pairs:
%   keepdata: Indicates whether the output table also should contain the
%   input data, i.e., by adding an extra column containing entropy values.
%   Defaults to true
%   parallel: Whether to run temporal entropy in parallel (true)
%
% Neurobiology Research Unit, 2023
% Please cite McCulloch, Olsen et al., 2023: "Navigating Chaos in
% Psychedelic Neuroimaging: A Rigorous Empirical Evaluation of the Entropic
% Brain Hypothesis" if you use CopBET in your studies. Please read the
% paper to get a notion of our recommendations regarding the use of the
% specific methodologies in the toolbox.

% ASO 9/3-2023

% Copyright (C) 2023 Anders Stevnhoved Olsen & Drummond E-Wen McCulloch
%
%    This program is free software: you can redistribute it and/or modify
%    it under the terms of the GNU General Public License as published by
%    the Free Software Foundation, either version 3 of the License, or
%    (at your option) any later version.
%
%    This program is distributed in the hope that it will be useful,
%    but WITHOUT ANY WARRANTY; without even the implied warranty of
%    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
%    GNU General Public License for more details.
%
%    You should have received a copy of the GNU General Public License
%    along with this program.  If not, see http://www.gnu.org/licenses/.

function out = CopBET_intranetwork_synchrony(in,atlas,varargin)
if nargin<2
    error('Please specify a 4D atlas')
end
[out,numworkers,in,NRUspecific] = CopBET_function_init(in,varargin);

%load data
disp('Beginning entropy calculations')

if ndims(atlas)~=4
    error('Please specify the atlas as a 4D matrix')
end

if numel(unique(atlas(:)))~=2
    error('please specify the atlas as a logical matrix')
end

parfor (ses = 1:height(in),numworkers)
    path = in{ses,1}{1};
    
    data_denoised = niftiread(path); %4D series
    
    if NRUspecific
        if ~cellfun(@isempty,regexp(path,'mr001'))
            data_denoised=NRUspecific_downsample_mr001data(data_denoised);
        end
    end
    datasz = size(data_denoised);
    atlassz = size(atlas);
    if ~all(datasz(1:3)==atlassz(1:3))
        error('data and atlas have different sizes')
    end
    
    variances = nan(size(atlas,4),datasz(4));
    
    for vol = 1:datasz(4)
        v = data_denoised(:,:,:,vol);
        v(v==0) = nan; % extra-ROI voxels would skew the variance
        
        for ROI = 1:size(atlas,4)
            variances(ROI,vol) = nanvar(v(atlas(:,:,:,ROI)));
        end
        
    end
    
    entropy_ROI = nan(size(atlas,4),1);
    for ROI = 1:size(atlas,4)
        [Prob,bins] = histcounts(variances(ROI,:),'Normalization','probability');
        entropy_ROI(ROI) = nansum(-Prob.*log(Prob));
    end
    entropy{ses} = entropy_ROI;
    sensible_data_check(entropy_ROI);
    
    disp(['Done with session ',num2str(ses),' of ',num2str(height(in))])
end

out.entropy = entropy';
end

%%

File: copbet_py/functions/CopBET_intranetwork_synchrony.py
"""
CopBET_intranetwork_synchrony
==============================
Copenhagen Brain Entropy Toolbox: Intranetwork synchrony.

Evaluates intranetwork synchrony as in Carhart-Harris et al., 2014.
For each time point and each ROI, the variance across voxels within the
ROI is computed. The Shannon entropy of the variance distribution over
time is returned per ROI.

This Python version accepts per-ROI voxelwise data rather than 4D NIfTI
files. Input is a list of sessions; each session is a list of 2D arrays
(one per ROI), where each array has shape (n_voxels_roi, T).

If ROI-level (single voxel per ROI) time series are provided, the
function automatically adds a small amount of noise to simulate voxel
spread, as this measure is only meaningful with multiple voxels per ROI.

Input
-----
sessions : list of list of np.ndarray
    sessions[i][roi] is shape (n_voxels, T) for ROI `roi` in session `i`.
    OR: list of np.ndarray each shape (T, N) — in this case each "voxel"
    is a single ROI signal and the function returns zeros (see note).

Returns
-------
entropy : list of np.ndarray, each shape (N_rois,)
    Per-ROI intranetwork synchrony entropy per session.

Reference
---------
Carhart-Harris et al., 2014.

Please cite McCulloch, Olsen et al., 2023 if you use CopBET.
"""

import numpy as np


def CopBET_intranetwork_synchrony(sessions, parcel_labels):
    """
    Compute intranetwork synchrony entropy.

    Parameters
    ----------
    sessions : list of (list of np.ndarray | np.ndarray)
        Each session is either:
        - A list of ROI arrays, where sessions[i][roi].shape = (n_voxels, T)
        - A 2D array of shape (T, N) — treated as N single-voxel ROIs

    Returns
    -------
    entropy : list of np.ndarray, each shape (N_rois,)
    """
    entropy = []

    labels = np.asarray(parcel_labels)
    unique_labels = np.unique(labels[labels != 0])
    N_rois = len(unique_labels)

    for ses, ts in enumerate(sessions):
        print(f"Processing session {ses + 1}/{len(sessions)}...", end='\r')
        ts = np.asarray(ts)

        T, N = ts.shape

        # Demean each voxel's time series
        ts = ts - ts.mean(axis=0)

        # Compute per-ROI, per-timepoint variance across voxels
        variances = np.zeros((T,N_rois))
        for i, roi in enumerate(unique_labels):
            roi_mask = labels == roi
            if np.sum(roi_mask) == 0:
                raise ValueError(f"ROI label {roi} has no voxels.")
            roi_ts = ts[:, roi_mask]  # (T, n_voxels_in_roi)
            variances[:,i] = np.nanvar(roi_ts, axis=1)  # (T,)

        # Shannon entropy of the variance distribution per ROI
        ent_roi = np.zeros(N_rois)
        for roi in range(N_rois):
            v = variances[:,roi]
            if np.all(v == 0):
                ent_roi[roi] = 0.0
                continue
            counts, _ = np.histogram(v, bins='auto')
            prob = counts / counts.sum()
            mask = prob > 0
            ent_roi[roi] = -np.sum(prob[mask] * np.log(prob[mask]))

        entropy.append(ent_roi)

    return entropy
